src/marqo/core/inference/inference_models/hugging_face_model.py

```diff
@@ -0,0 +1,249 @@
+            max_length=self.model_properties.tokens,
+            return_tensors="pt"
+        ).to(self.device)
+
+        with torch.no_grad():
+            model_output = self._model(**tokenized_texts)
+
+        attention_mask = tokenized_texts['attention_mask']
+
+        embeddings = self._pooling_func(model_output, attention_mask)
+
+        if normalize:
+            embeddings = F.normalize(embeddings, p=2, dim=1)
+
+        return self._convert_output(embeddings)
+
+    def _convert_output(self, output):
+        if self.device == 'cpu':
+            return output.numpy()
+        elif self.device.startswith('cuda'):
+            return output.cpu().numpy()
+
+    @staticmethod
+    def _average_pool_func(model_output, attention_mask):
+        """A pooling function that averages the hidden states of the model."""
+        last_hidden = model_output.last_hidden_state.masked_fill(~attention_mask[..., None].bool(), 0.0)
+        return last_hidden.sum(dim=1) / attention_mask.sum(dim=1)[..., None]
+
+    @staticmethod
+    def _cls_pool_func(model_output, attention_mask=None):
+        """A pooling function that extracts the CLS token from the model."""
+        return model_output[0][:, 0]
+
+    @staticmethod
+    def extract_huggingface_archive(path: str) -> str:
+        '''
+            This function takes the path as input. The path can must be a string that can be:
+            1. A downloaded archive file. This function will extract the model from the archive return the directory path.
+            2. A repo_id in huggingface. This function will return the input string directly.
+
+            path: the downloaded model archive path or a repo_id in huggingface
+        Returns:
+            The directory path to the model or the repo_id in huggingface
+        '''
+        if os.path.isfile(path):
+            # if it's a file, check if it's a compressed file
+            base, ext = os.path.splitext(path)
+            if ext in ['.bin', '.pt']:
+                raise InvalidModelPropertiesError(f"Marqo does not support loading Hugging Face SBERT models from the provided single `{ext}` file. "
+                                                  "Please try to wrap the model in a Hugging Face archive file and try again. ")
+            try:
+                # create a new directory with the same name as the file
+                new_dir = base
+                os.makedirs(new_dir, exist_ok=True)
+
+                # extract the compressed file
+                # If the target directory already exists, it will be overwritten by default without warning.
+                if ext == '.zip':
+                    with zipfile.ZipFile(path, 'r') as zip_ref:
+                        zip_ref.extractall(new_dir)
+                else:
+                    with tarfile.open(path, 'r') as tar_ref:
+                        tar_ref.extractall(new_dir)
+                # return the path to the new directory
+                return new_dir
```

Comment:
Do we need to keep the extracted files after the model is loaded?

Reply:
Yes. The model can be ejected and reloaded later. Keeping the extracted files so we don't have to redownload and reextract.